pkg/csi/service/common/authmanager.go

```diff
@@ -0,0 +1,158 @@
+package common
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	cnsvsphere "sigs.k8s.io/vsphere-csi-driver/pkg/common/cns-lib/vsphere"
+	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/logger"
+)
+
+type AuthorizationService interface {
+
+	// GetDatastoreIgnoreMapForBlockVolumes returns a map which maps datastore url to datastore info which user does not have Datastore.FileManagement privilege
+	// returns an error if not able to retrieve the map
+	GetDatastoreIgnoreMapForBlockVolumes(ctx context.Context) (map[string]*cnsvsphere.DatastoreInfo, error)
+}
+
+type AuthManager struct {
+	// map the datastore url to datastore info which need to be ignored when invoking CNS CreateVolume API
+	datastoreIgnoreMapForBlockVolumes map[string]*cnsvsphere.DatastoreInfo
+	// this mutex is to make sure the update for datatoreIgnoreMap is mutually exclusive
+	mutex sync.Mutex
+	// vCenter Instance
+	vcenter *cnsvsphere.VirtualCenter
+}
+
+// onceForAuthorizationService is used for initializing the AuthorizationService singleton.
+var onceForAuthorizationService sync.Once
+
+// authManagerIntance is instance of authManager and implements interface for AuthorizationService
+var authManagerInstance *AuthManager
+
+// GetAuthorizationService returns the singleton AuthorizationService
+func GetAuthorizationService(ctx context.Context, vc *cnsvsphere.VirtualCenter) (AuthorizationService, error) {
+	log := logger.GetLogger(ctx)
+	onceForAuthorizationService.Do(func() {
+		log.Info("Initializing authorization service...")
+
+		authManagerInstance = &AuthManager{
+
+			datastoreIgnoreMapForBlockVolumes: make(map[string]*cnsvsphere.DatastoreInfo),
+			mutex:                             sync.Mutex{},
+			vcenter:                           vc,
+		}
+		RefreshDatastoreIgnoreMap(authManagerInstance)
+	})
+
+	log.Info("authorization service initialized")
+	return authManagerInstance, nil
+}
+
+func (authManager *AuthManager) GetDatastoreIgnoreMapForBlockVolumes(ctx context.Context) (map[string]*cnsvsphere.DatastoreInfo, error) {
+	authManager.mutex.Lock()
+	defer authManager.mutex.Unlock()
+	var datastoreIgnoreMapForBlockVolumes map[string]*cnsvsphere.DatastoreInfo
+	for dsURL, dsInfo := range authManager.datastoreIgnoreMapForBlockVolumes {
+		datastoreIgnoreMapForBlockVolumes[dsURL] = dsInfo
+	}
+	return datastoreIgnoreMapForBlockVolumes, nil
+}
+
+// RefreshDatastoreIgnoreMap scans all datastores in vCenter to check privileges, and compute the DatastoreIgnoreMap
+func RefreshDatastoreIgnoreMap(authManager *AuthManager) {
```

Comment:
This can be a private method in AuthManager.

Reply:
Done.